### PR TITLE
LibGfx + LibWeb: Implement GIF animation

### PR DESCRIPTION
**Pull request by peterdn**
This PR:
- LibGfx: Add methods applicable to animated images to ImageDecoder and ImageDecoderPlugin interfaces.
- LibGfx: Implement animated image methods for GIFImageDecoderPlugin and optimise LZWDecoder a little.
- LibWeb: Add support for animated images to HTMLImageElement.

Upshot: animated GIFs are now displayed in the browser. Includes support for variable delays between frames, and single-loop GIFs.

SerenityGIFs

Things to pay particular attention to:
- Animation methods added to ImageDecoder{Plugin} interfaces: felt like the right place to put them but happy to consider alternative approaches.
- Decoding is done lazily but is quite slow and memory-intensive. I intend to optimise in a follow-up PR but can do it here if requested.

### CODE REVIEW

File: Libraries/LibGfx/ImageDecoder.h [hunk 1]
```diff
@@ -26,15 +26,20 @@
...
+#include <AK/RefPtr.h>
 #include <LibGfx/Size.h>
 
 namespace Gfx {
 
 class Bitmap;
 
+struct ImageFrameDescriptor {
+    RefPtr<Bitmap> image;
+    int duration;
```
Comment (awesomekling) [inline Libraries/LibGfx/ImageDecoder.h:40]: Please initialize primitive variables when declaring them.

File: Libraries/LibWeb/DOM/HTMLImageElement.cpp [hunk 2]
```diff
@@ -37,6 +38,9 @@ namespace Web {
 
 HTMLImageElement::HTMLImageElement(Document& document, const FlyString& tag_name)
     : HTMLElement(document, tag_name)
+    , m_current_frame_index(0)
+    , m_loops_completed(0)
```
Comment (awesomekling) [inline Libraries/LibWeb/DOM/HTMLImageElement.cpp:42]: Please initialize these at the declaration instead :)

**Comment by peterdn** [comment]
Thanks @awesomekling!

> Feature request after this: interlacing
Could be fun :-)

**Comment by awesomekling** [comment]
Why is `loop_count()` an `int`? Can it be negative?

**Comment by peterdn** [comment]
> Why is `loop_count()` an `int`? Can it be negative?

Hmm, probably not. Think I was leaving my options open in case -1 was needed to mean infinite, but 0 does instead. I can make it a `u32` instead.

**Comment by awesomekling** [comment]
Or `size_t` for symmetry with `frame_count()` :)

**Comment by linusg** [comment]
Commits probably need some squashing now :)

https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy

> - Squash your commits when making revisions after a patch review.

**Comment by awesomekling** [comment]
Grand!

**Pull request merged**